File: src/sources/console/hook.ts
import type { LogStore } from '../../core/store.js';
import {
	CONSOLE_METHODS,
	ConsoleRecorder,
	type ConsoleMethod,
	type RecorderOptions
} from './recorder.js';

export interface HookConsoleOptions extends Partial<RecorderOptions> {
	/** The methods to hook. Defaults to every method in `CONSOLE_METHODS`. */
	methods?: readonly ConsoleMethod[];
	/** Whether the original method still runs, so messages keep reaching the browser console. */
	passthrough?: boolean;
}

/** An object with the console methods lognal records. */
export type LognalConsole = { [Method in ConsoleMethod]: (...args: unknown[]) => void };

type ConsoleLike = Record<string, unknown>;
type StackCapture = (target: object, constructorOpt?: (...args: never[]) => unknown) => void;

const FRAME_V8 = /^\s+at\s/;
const FRAME_OTHER = /@/;

/** Removes the lines before the first stack frame, such as V8's `Error` header. */
const framesOnly = (stack: string): string[] => {
	const lines = stack.split('\n');
	const first = lines.findIndex((line) => FRAME_V8.test(line) || FRAME_OTHER.test(line));

	return first < 0 ? [] : lines.slice(first);
};

/**
 * Returns the stack trace of the code that called `wrapper`, without the frames of lognal
 * itself. V8 can leave those frames out on its own; elsewhere the first frame is dropped.
 */
const captureStack = (wrapper: (...args: never[]) => unknown): string => {
	const captureStackTrace = (Error as unknown as { captureStackTrace?: StackCapture })
		.captureStackTrace;

	if (typeof captureStackTrace === 'function') {
		const holder: { stack?: string } = {};

		captureStackTrace(holder, wrapper);

		return framesOnly(holder.stack ?? '').join('\n');
	}

	return framesOnly(new Error().stack ?? '')
		.slice(2)
		.join('\n');
};

/**
 * Replaces the methods of a console so every call is also recorded in a store.
 *
 * Returns a function that restores the original methods. If another script wrapped a method
 * after lognal did, that method is left in place and lognal's wrapper stops recording.
 *
 * A call is recorded before the original method runs, and an error while recording never
 * reaches the page.
 */
export const hookConsole = (
	target: Console,
	store: LogStore,
	options: HookConsoleOptions = {}
): (() => void) => {
	const { methods = CONSOLE_METHODS, passthrough = true, ...recorderOptions } = options;
	const recorder = new ConsoleRecorder(store, recorderOptions);
	const console = target as unknown as ConsoleLike;
	const installed = new Map<ConsoleMethod, { original: unknown; wrapper: unknown }>();
	let active = true;
	let recording = false;

	for (const method of methods) {
		if (installed.has(method)) {
			continue;
		}

		const original = console[method];
		const wrapper = function (this: unknown, ...args: unknown[]): unknown {
			if (active && !recording) {
				recording = true;

				try {
					recorder.record(method, args, method === 'trace' ? captureStack(wrapper) : undefined);
				} catch {
					// Recording must never break the page that logged the message.
				} finally {
					recording = false;
				}
			}

			if (passthrough && typeof original === 'function') {
				return original.apply(this, args);
			}

			return undefined;
		};

		console[method] = wrapper;
		installed.set(method, { original, wrapper });
	}

	return () => {
		if (!active) {
			return;
		}

		active = false;

		for (const [method, { original, wrapper }] of installed) {
			if (console[method] === wrapper) {
				console[method] = original;
			}
		}
	};
};

/**
 * Creates an object with the console methods that records into a store without touching the
 * global console. Use it to write to a viewer from your own code.
 */
export const createConsole = (
	store: LogStore,
	options: Partial<RecorderOptions> = {}
): LognalConsole => {
	const recorder = new ConsoleRecorder(store, options);
	const result = {} as LognalConsole;

	for (const method of CONSOLE_METHODS) {
		const call = (...args: unknown[]): void => {
			recorder.record(method, args, method === 'trace' ? captureStack(call) : undefined);
		};

		result[method] = call;
	}

	return result;
};

File: packages/flutter/lib/src/sources/console/hook.dart
import 'dart:async';

import 'package:flutter/foundation.dart';
import 'package:lognal/src/core/store.dart';
import 'package:lognal/src/core/types.dart';
import 'package:lognal/src/sources/console/recorder.dart';

/// How much of what the application already prints reaches the viewer.
class HookOptions {
  /// Creates the options.
  const HookOptions({
    this.recorder = defaultRecorderOptions,
    this.passthrough = true,
    this.level = LogLevel.log,
  });

  /// What a recorded call captures. Only [hookFlutterErrors] captures values;
  /// the other two hooks receive text that is already formatted.
  final RecorderOptions recorder;

  /// Whether the original output still runs, so messages keep reaching the
  /// terminal and the IDE.
  final bool passthrough;

  /// The level every captured line is given. [hookFlutterErrors] ignores it,
  /// because what it captures is an error.
  final LogLevel level;
}

/// Records everything `debugPrint` writes into a store, for as long as the
/// returned function has not been called.
///
/// This is Dart's answer to hooking the console: `debugPrint` is a variable
/// holding a function, so it can be replaced and put back, and it is what
/// Flutter itself prints through. If another package wrapped it after lognal
/// did, that wrapper is left in place and lognal's stops recording, the same way
/// the JavaScript package leaves a console method somebody else replaced.
///
/// `print` cannot be replaced like this, because it belongs to the zone rather
/// than to a variable. Use [runZonedWithLognal] for that.
void Function() hookDebugPrint(LogStore store, {HookOptions options = const HookOptions()}) {
  final DebugPrintCallback original = debugPrint;
  bool active = true;
  bool recording = false;

  void wrapper(String? message, {int? wrapWidth}) {
    if (active && !recording && message != null) {
      recording = true;

      try {
        store.writeLines(message, WriteOptions(level: options.level));
      } catch (_) {
        // Recording must never break the code that logged the message.
      } finally {
        recording = false;
      }
    }

    if (options.passthrough) {
      original(message, wrapWidth: wrapWidth);
    }
  }

  debugPrint = wrapper;

  return () {
    if (!active) {
      return;
    }

    active = false;

    if (debugPrint == wrapper) {
      debugPrint = original;
    }
  };
}

/// Runs [body] in a zone whose `print` writes into a store.
///
/// `print` is resolved through the current zone rather than through a variable,
/// so it is captured by running the code inside a zone rather than by replacing
/// anything. Wrap `runApp` in this to catch what the application prints.
R runZonedWithLognal<R>(
  LogStore store,
  R Function() body, {
  HookOptions options = const HookOptions(),
}) {
  return runZoned(
    body,
    zoneSpecification: ZoneSpecification(
      print: (Zone self, ZoneDelegate parent, Zone zone, String message) {
        try {
          store.writeLines(message, WriteOptions(level: options.level));
        } catch (_) {
          // Recording must never break the code that printed the message.
        }

        if (options.passthrough) {
          parent.print(zone, message);
        }
      },
    ),
  );
}

/// Records the errors the framework reports into a store, for as long as the
/// returned function has not been called.
///
/// A widget that throws during a build, a layout that overflows and a failed
/// image load all arrive here, which is the part of a Flutter application's
/// output a developer most wants in front of them.
void Function() hookFlutterErrors(LogStore store, {HookOptions options = const HookOptions()}) {
  final ConsoleRecorder recorder = ConsoleRecorder(store, options: options.recorder);
  final FlutterExceptionHandler? original = FlutterError.onError;
  bool active = true;

  void wrapper(FlutterErrorDetails details) {
    if (active) {
      try {
        recorder.record(ConsoleMethod.error, <Object?>[
          details.exception,
          if (details.context != null) '${details.context}',
        ], stack: details.stack);
      } catch (_) {
        // Recording must never swallow the error it was reporting.
      }
    }

    if (options.passthrough) {
      (original ?? FlutterError.presentError)(details);
    }
  }

  FlutterError.onError = wrapper;

  return () {
    if (!active) {
      return;
    }

    active = false;

    if (FlutterError.onError == wrapper) {
      FlutterError.onError = original;
    }
  };
}
